modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java

```diff
@@ -0,0 +1,269 @@
+        if (!busyLock.enterBusy()) {
+            return CompletableFuture.failedFuture(new IgniteSqlException("Session is closed."));
+        }
+
+        QueryContext ctx = QueryContext.of(transaction, new QueryTimeout(timeout, TimeUnit.NANOSECONDS));
+
+        try {
+            List<CompletableFuture<AsyncSqlCursor<List<Object>>>> futs = qryProc.queryAsync(ctx, schema, query, arguments);
+
+            futsToClose.addAll(futs);
+
+            if (futs.size() != 1) {
+                return CompletableFuture.failedFuture(new IgniteSqlException("Multiple statements aren't allowed."));
+            }
+
+            return futs.get(0).thenCompose(cur -> {
+                        futsToClose.remove(futs.get(0));
+
+                        if (!busyLock.enterBusy()) {
+                            return cur.closeAsync()
+                                    .thenCompose((v) -> CompletableFuture.failedFuture(new IgniteSqlException("Session is closed")));
+                        }
+
+                        try {
+                            cursToClose.add(cur);
+
+                            return cur.requestNextAsync(pageSize)
+                                    .thenApply(
+                                            batchRes -> new AsyncResultSetImpl(
+                                                    cur,
+                                                    batchRes,
+                                                    pageSize,
+                                                    () -> cursToClose.remove(cur)
+                                            )
+                                    );
+                        } finally {
+                            busyLock.leaveBusy();
+                        }
+                    }
+            );
+        } catch (Exception e) {
+            return CompletableFuture.failedFuture(e);
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<AsyncResultSet> executeAsync(@Nullable Transaction transaction, Statement statement) {
+        throw new UnsupportedOperationException("Not implemented yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<Integer> executeBatchAsync(@Nullable Transaction transaction, String query, BatchedArguments batch) {
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<Integer> executeBatchAsync(@Nullable Transaction transaction, Statement statement, BatchedArguments batch) {
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    public CompletableFuture<Void> closeAsync() {
+        return CompletableFuture
+                .runAsync(busyLock::block)
+                .thenCompose(
+                        v0 -> {
+                            futsToClose.forEach(f -> f.cancel(true));
```

Comment:
cancellation of the future doesn't cancel the whole computation chain started. So we'll get a bunch of opened cursors here